fs: Fix default params for fs.write

papandreou opened the pull request:
##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines
##### Affected core subsystem(s)

fs
##### Description of change

Add support for `fs.write(fd, buffer, cb)` and `fs.write(fd, buffer, offset, cb)` as documented at https://nodejs.org/api/fs.html#fs_fs_write_fd_data_position_encoding_callback (it says that `data` can be either a string or a Buffer).

Update: Extended to apply to `fs.writeSync` as well, which turned out to have the same issue (#7879).

mscdex commented on test/sequential/test-fs-write-buffer-without-length.js:
```diff
@@ -0,0 +1,27 @@
+  fs.write(fd,
+           Buffer.from('hello'),
+           3,
+           common.mustCall(function(err, written) {
```
This looks a bit hectic. What about moving this `common.mustCall(...)` to a variable and passing that as the fourth argument instead? That way the `fs.write()` fits all on one line.

papandreou replied:
Seems like I copy/pasted from a very old version of the https://github.com/nodejs/node/blob/master/test/parallel/test-fs-buffer.js test. Will try to fix it up.

papandreou replied:
https://github.com/nodejs/node/blob/master/test/parallel/test-fs-write-buffer.js, actually, and that's still just as messy :)

mscdex replied:
Perhaps, but let's not continue that trend ;-)

papandreou replied:
Agreed. Fixed in https://github.com/nodejs/node/pull/7856/commits/c9d7a7744f56ffc468da53937dc4a84d518d2693

In both tests I had to keep the definition of `cb` inside the other callback because it references `fd` in `fs.closeSync(fd);`.

mscdex commented on test/sequential/test-fs-write-buffer-without-offset-and-length.js:
```diff
@@ -0,0 +1,27 @@
+
+  fs.write(fd,
+           expected,
+           common.mustCall(function(err, written) {
```
Same here.

mscdex commented on test/sequential/test-fs-write-buffer-without-length.js:
```diff
@@ -0,0 +1,26 @@
+fs.open(filename, 'w', 0o644, common.mustCall(function(err, fd) {
+  if (err) throw err;
+
+  var cb = common.mustCall(function(err, written) {
```
minor nit: use `const` instead of `var` here and the other instance below.

papandreou replied:
Thanks, fixed.

mscdex commented on test/sequential/test-fs-write-buffer-without-offset-and-length.js:
```diff
@@ -0,0 +1,27 @@
+fs.open(filename, 'w', 0o644, common.mustCall(function(err, fd) {
+  if (err) throw err;
+
+  var cb = common.mustCall(function(err, written) {
```
Ditto here and below.

papandreou replied:
Fixed.

mscdex commented:
Any reason the tests are in `sequential/` instead of `parallel/`?

papandreou commented:
@mscdex Not a good one, no. I just misunderstood the sequential vs. parallel classification. Moved them to `parallel` now :)

thefourtheye commented on test/parallel/test-fs-write-buffer-without-length.js:
```diff
@@ -0,0 +1,26 @@
+common.refreshTmpDir();
+
+fs.open(filename, 'w', 0o644, common.mustCall(function(err, fd) {
+  if (err) throw err;
```
You can use `assert.ifError`

thefourtheye commented on test/parallel/test-fs-write-buffer-without-length.js:
```diff
@@ -0,0 +1,26 @@
+  const cb = common.mustCall(function(err, written) {
+    if (err) throw err;
+
+    assert.equal(2, written);
```
Use `strictEqual`

thefourtheye commented on test/parallel/test-fs-write-buffer-without-offset-and-length.js:
```diff
@@ -0,0 +1,27 @@
+
+    const found = fs.readFileSync(filename, 'utf8');
+    assert.deepStrictEqual(expected.toString(), found);
+    fs.unlinkSync(filename);
```
This is not necessary, as the temp directory will be cleaned anyway.

thefourtheye commented on test/parallel/test-fs-write-buffer-without-offset-and-length.js:
```diff
@@ -0,0 +1,27 @@
+const Buffer = require('buffer').Buffer;
+const fs = require('fs');
+const filename = path.join(common.tmpDir, 'write.txt');
+const expected = Buffer.from('hello');
```
This can be a string itself.

papandreou commented:
Fixed the nits, including the test that I copied from.

Also fixed `fs.writeSync` in the same way (came up in #7879).

yorkie commented on test/parallel/test-fs-write-buffer.js:
```diff
@@ -10,21 +10,17 @@ const expected = Buffer.from('hello');
 
-             var found = fs.readFileSync(filename, 'utf8');
-             assert.deepStrictEqual(expected.toString(), found);
-             fs.unlinkSync(filename);
```
Why removing this line?

papandreou replied:
@thefourtheye suggested that: https://github.com/nodejs/node/pull/7856#discussion_r72006821

yorkie commented on test/parallel/test-fs-write-sync.js:
```diff
@@ -1,23 +1,23 @@
-assert.ok(written > 3);
+const bar = 'bár';
+const written2 = fs.writeSync(fd, Buffer.from(bar), 0, Buffer.byteLength(bar));
+assert.ok(written2 > 3);
```
```js
{
  let written = ...
  // here write different writtens
}
```
How about using `let` in block to do scoping here?

papandreou replied:
Changed it to use `let` in 8dc500710, although without the extra block. I hope that's OK.

cjihrig commented on test/parallel/test-fs-write-buffer-without-length.js:
```diff
@@ -0,0 +1,25 @@
+const common = require('../common');
+const assert = require('assert');
+const path = require('path');
+const Buffer = require('buffer').Buffer;
```
You shouldn't need to require `Buffer` here and in the other tests.

papandreou replied:
Fixed in ff232027.

cjihrig commented on test/parallel/test-fs-write-sync-without-length.js:
```diff
@@ -0,0 +1,23 @@
+const assert = require('assert');
+const path = require('path');
+const fs = require('fs');
+const fn = path.join(common.tmpDir, 'write.txt');
```
Could you use a more descriptive variable name here please.

papandreou replied:
Fixed in 31793894.

cjihrig commented on test/parallel/test-fs-write-sync-without-length.js:
```diff
@@ -0,0 +1,23 @@
+assert.ok(written2 > 3);
+fs.closeSync(fd);
+
+assert.equal(fs.readFileSync(fn), 'foobár');
```
`assert.strictEqual()` please.

papandreou replied:
Fixed in eb6c4e866.

cjihrig commented on test/parallel/test-fs-write-sync-without-length.js:
```diff
@@ -0,0 +1,23 @@
+
+common.refreshTmpDir();
+
+const foo = 'foo';
```
You could just drop the `foo` and `bar` variables and use their values inline.

papandreou replied:
Fixed in 6322f2c33.

cjihrig commented on test/parallel/test-fs-write-sync-without-length.js:
```diff
@@ -0,0 +1,23 @@
+const foo = 'foo';
+const fd = fs.openSync(fn, 'w');
+
+const written = fs.writeSync(fd, '');
```
You could change this to `let`, and drop the `written2` variable.

papandreou replied:
Fixed in 8dc50071.

cjihrig commented on test/parallel/test-fs-write-sync-without-offset-and-length.js:
```diff
@@ -0,0 +1,23 @@
+'use strict';
```
Most of the same comments apply to this test. You could probably combine the two tests in one file, separated by block scopes.

papandreou replied:
Ported the changes to test-fs-write-sync-without-offset-and-length.js as well now.

Not sure I'd be comfortable combining them into one (and especially not their async counterparts) -- isn't it better to have more (simple) tests that can fail individually?

cjihrig replied:
Please see https://github.com/nodejs/node/blob/master/test/parallel/test-fs-mkdir.js as an example of what I mean. In some cases, it's overkill to have a bunch of files doing almost exactly the same thing.

papandreou replied:
Okay, I hear you. Fixed in 15d9002.

Did the same thing for the `fs.write` tests in 628ee89d. Let me know if you'd like that included as well.

jasnell commented:
Question for @nodejs/ctc : is this a patch or a semver-minor?

trevnorris commented on lib/fs.js:
```diff
@@ -735,6 +735,12 @@ fs.write = function(fd, buffer, offset, length, position, callback) {
+    }
+    if (typeof length !== 'number') {
+      length = buffer.length - offset;
+    }
```
The comment at the top of the function does not match public documentation. Mind fixing that.

papandreou replied:
Sure thing, fixed in 4062d4cf. I changed both the comment and the documentation to match the implementation.

trevnorris commented:
While doing the review I realized that our documented API doesn't come close to following current implementation. For example, `src/node_file.cc:node::WriteBuffer()` still has all the same throw logic as before. While `src/node_file.cc:node::WriteString()` relies on coercion rules from external calls to `ParseEncoding()`, etc.

There is a deeper issue that needs to be fixed. This PR does help bring it closer to compliance to public docs, but is only a superficial fix. I'm fine with this landing, but there should be a follow-up PR that fixes things all the way down.

@jasnell Technically a patch since it aligns implementation closer to documentation, and it shouldn't introduce any regressions.

jasnell commented:
@trevnorris +1

papandreou commented:
Squashed and rebased on master as conflicts had been introduced.

The code that works out the default values of the default parameters for `fs.write` got a bit more complex because the line `callback = makeCallback(arguments[arguments.length - 1]);` was removed from master in the mean time.

I believe this work is finished and ready to land now, but let me know if you have any feedback, especially about the code that sets the default parameters and finds the callback.

trevnorris commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/3626/

jasnell commented:
LGTM if CI is green.

papandreou commented:
Fixed the lint error and rebased.

trevnorris commented:
@jasnell i only see AIX and those all look flaky. I'm cool w/ this landing.

jasnell commented:
@trevnorris should I count that as a LGTM?

jasnell commented:
@mscdex ... does this LGTY?

mscdex commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/3744/

mscdex commented:
LGTM

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/3790/

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/3807/

jasnell commented:
There was some odd red in the that last one... trying yet again to see if it was a flaky fluke: https://ci.nodejs.org/job/node-test-pull-request/3825/

jasnell commented:
Ok, this appears to be consistently failing on AIX for some reason.

papandreou commented:
@jasnell Seems like it worked this time, but now it fails on ppc?

(Not sure if there's anything I can do)

jasnell commented:
@papandreou ... those might be some flaky failures. I'll take a look again later on today.

lpinca commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/4275/

gibfahn commented:
This AIX failure seems consistent (and I haven't seen it on test runs for any other PRs) so it needs to be looked into.

Failure for posterity:
```
not ok 385 parallel/test-fs-write-buffer
# 
# assert.js:85
#   throw new assert.AssertionError({
#   ^
# AssertionError: 'hello' deepStrictEqual ''
#     at /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/test/parallel/test-fs-write-buffer.js:103:14
#     at /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/test/common.js:421:15
#     at FSReqWrap.wrapper [as oncomplete] (fs.js:672:5)
  ---
  duration_ms: 0.336
  ...
```

mhdawson commented:
Talked to @gibfahn he should be starting to look at it shortly.

Akpotohwo commented:
I have just run tests once on AIX and while parallel/test-fs-write-buffer did not fail, parallel/test-fs-readfile-tostring-fail did ... I will now run both in a loop and see if they are intermittent or consistent.

Akpotohwo commented:
They both seem to be consistent for me. I will investigate further

Akpotohwo commented on test/parallel/test-fs-write-buffer.js:
```diff
@@ -2,29 +2,107 @@
+// fs.write with the offset passed as undefined followed by the callback:
+{
+  const filename = path.join(common.tmpDir, 'write3.txt');
+  fs.open(filename, 'w', 0o644, common.mustCall(function(err, fd) {
```
Shouldn't these last 3 subtests all write to different files and not just `write3.txt`?
It seems what's happening here is the fs calls are running asynchronously; and since they are pointing to the same file, it is overwritten before it is read and compared.
I'd suggest changing the extra test filenames to `write4.txt` and `write5.txt`

papandreou replied:
Oh, man, you're right. I've fixed that in 747fdfb -- could someone please start a new CI build?

sam-github commented on doc/api/fs.md:
```diff
@@ -1577,13 +1577,13 @@ _Note: [`fs.watch()`][] is more efficient than `fs.watchFile` and
 
 * `fd` {Integer}
-* `buffer` {String | Buffer}
+* `buffer` {Buffer}
```
From the doc changes, this appears to be semver major, in that it removes the `fs.write(fd, string, offset, length, position, callback)` as a supported use. Was that usage not implemented, so its aligning docs with actual implementation?

papandreou replied:
Yes, it wasn't ever implemented. The code ends up either [calling binding.writeBuffer](https://github.com/nodejs/node/blob/e8fadd8da5a2d073815ba25a3eea470285a0b178/lib/fs.js#L684) or [binding.writeString](https://github.com/nodejs/node/blob/e8fadd8da5a2d073815ba25a3eea470285a0b178/lib/fs.js#L699) based on whether the second parameter is a buffer, and those functions aren't polymorphic.

sam-github commented on doc/api/fs.md:
```diff
@@ -1608,19 +1608,19 @@ On Linux, positional writes don't work when the file is opened in append mode.
 the end of the file.
 
-## fs.write(fd, data[, position[, encoding]], callback)
+## fs.write(fd, string[, position[, encoding]], callback)
```
Is this syntax correct? Is it really impossible to specify encoding unless you also specify position? So, to write using an encoding at the current fd position, I would have to do `fs.write(fd, string, nil, encoding)`, specifying a position that has typeof != number?

papandreou replied:
Correct! It does seem a bit convoluted, yes. This time around I only meant to fix the bugs and make sure the documentation is correct, though :)

gibfahn commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/4733/

papandreou commented:
@gibfahn Thanks, looks like that helped. I assume that FreeBSD failure is unrelated?

gibfahn commented:
@papandreou Almost certainly. I'll stress run the test on AIX tomorrow just to make sure everything's okay, and then we should be good to go.

sam-github commented:
Fixes https://github.com/nodejs/node/pull/9504

Fixes https://github.com/nodejs/node/issues/9479

sam-github commented:
@gibfahn LGTM, LGTY?

gibfahn commented:
LGTM pending https://ci.nodejs.org/job/node-stress-single-test/1060/ (a.k.a. https://ci.nodejs.org/job/node-stress-single-test/nodes=aix61-ppc64/1060/console)

sam-github commented:
@gibfahn you aborted the test runs you reference above?

gibfahn commented:
No failures in 4479 runs, LGTM.

Thanks for the ping @sam-github

sam-github commented:
@papandreou I assume the second commit is a fixup on the first, and should be squashed into it, because its commit message doesn't meet the convention, and it seems to be an adjustment to the first commit?

gibfahn commented:
@sam-github Yeah, you couldn't merge without squashing them, otherwise the tests would be broken on the first commit.

papandreou commented:
@sam-github @gibfahn Good point -- squashed and force-pushed.

gibfahn commented:
CI is green, @sam-github are you landing?

sam-github commented:
I will land this

sam-github commented:
Landed in 0b5191f, thanks @papandreou

TheAlphaNerd commented:
added lts labels. Is this going to be reverted as per #10242?

gibfahn commented:
It sounds like this is a minor behavioural change, where the previous code didn't do exactly what the docs said. Given that this was [reported as an issue in v6](https://github.com/nodejs/node/issues/7879), it might be worth backporting.

Thoughts @nodejs/lts ?

sam-github commented:
1. This adds support for new arg combinations, shouldn't it be tagged semver-minor?
2. @papandreou Can you think of any negative side-effects? Is this purely additive, or does it change an existing behaviour? An existing behaviour that is useful, I mean, changing the existing behaviour of writing absolutely none of the buffer doesn't count.

papandreou commented:
@sam-github, no, I can't think of any negative side effects of backporting, especially now that it has been on v7 for a while without causing any problems. To my knowledge the change in behavior has only been brought up once (#10242), and that was resolved amicably.

Based on the number of reports of the original problem, I agree that it seems to be something that comes up often enough that we could prevent a few wtf experiences for 6.x users by backporting it.

sam-github commented:
Above makes sense to me, @papandreou, thanks. Do you have time to do the backport this by opening a PR against the v6.x-staging branch?

papandreou commented:
@sam-github, sure! https://github.com/nodejs/node/pull/13179
